### PR TITLE
[Blazor] [WebView] Avoid Base64 encoding because Json Serializer should do that

### LINKED ISSUE #43867: Blazor Hybrid - Performance UpdateDisplayAsync
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We have a component with a high frequency of updates in case the user is scrolling within the component.
This feels a bit laggy, and we have looked up the situation with dottrace.

We optimized our implementation already, but we see an high CPU usage within the Internals of the WebView Component of Blazor. Especially the Method "UpdateDisplayAsync".

image

I can see here two things, one is the Serialize Method of the IpcCommon class, and also the PostWebMessageAsString Method of the WebView2 component.

Do you have any advise we can improve the situation here?

### Expected Behavior

Fast Performance :-)

### Steps To Reproduce

_No response_

### Exceptions (if any)

_No response_

### .NET Version

6.0.8

### Anything else?

Blazor Hybrid with WPF Integration.
```
<PackageReference Include="Microsoft.AspNetCore.Components.WebView.Wpf" Version="6.0.486" />
<PackageReference Include="Microsoft.AspNetCore.Components.Forms" Version="6.0.8" />
<PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="6.0.8" />
```

### PR DESCRIPTION
**Pull request by softwaretirol**
# [Blazor] [WebView] Avoid Base64 encoding because Json Serializer should do that

Currently WebView is converting the payload to Base64 and handing over it to the internal Json serializer. In case the payload is quite large, the serialization takes a large amount of time, because the serializer is handling the string value as "normal" and tries to escape the content of the string. This PR is handing over a byte[] to the serializer, so the serializer converters it to Base64 without further check.

Fixes #43867

### CODE REVIEW

File: src/Components/WebView/WebView/src/IpcSender.cs [hunk 1]
```diff
@@ -29,7 +29,12 @@ public void ApplyRenderBatch(long batchId, RenderBatch renderBatch)
         {
             renderBatchWriter.Write(in renderBatch);
         }
-        var message = IpcCommon.Serialize(IpcCommon.OutgoingMessageType.RenderBatch, batchId, Convert.ToBase64String(arrayBuilder.Buffer, 0, arrayBuilder.Count));
+
+        var data = arrayBuilder.Buffer.AsSpan(0, arrayBuilder.Count).ToArray();
+        var message = IpcCommon.Serialize(IpcCommon.OutgoingMessageType.RenderBatch,
+            batchId,
+            data);
```
Comment (javiercn) [inline src/Components/WebView/WebView/src/IpcSender.cs:36]: I do not think that this by itself will be enough, the JS code likely has to react to this.

Comment (softwaretirol) [reply src/Components/WebView/WebView/src/IpcSender.cs:36]: It is binary the same, for the consumer side (JS side) it looks the same. So there is no JS adaption needed?
If the Base64 encoding is done before calling Serialize, or the Serializer itself is doing that for us makes no difference.

File: src/Components/WebView/WebView/src/IpcSender.cs [hunk 2]
```diff
@@ -29,7 +29,12 @@ public void ApplyRenderBatch(long batchId, RenderBatch renderBatch)
         {
             renderBatchWriter.Write(in renderBatch);
         }
-        var message = IpcCommon.Serialize(IpcCommon.OutgoingMessageType.RenderBatch, batchId, Convert.ToBase64String(arrayBuilder.Buffer, 0, arrayBuilder.Count));
+
+        var data = arrayBuilder.Buffer.AsSpan(0, arrayBuilder.Count).ToArray();
```
Comment (TanayParikh) [inline src/Components/WebView/WebView/src/IpcSender.cs:33]: By doing `ToArray` here, are we potentially allocating additional memory, which was avoided when we simply converted the array builder span to a base 64 string?

Comment (softwaretirol) [reply src/Components/WebView/WebView/src/IpcSender.cs:33]: If we convert the span to a base64 string, a string will be allocated too. In addition, the serializer has to intgrate the string again. 

I would argue here that the amount of memory might be higher with the string, because the string encoded as base 64 is much longer?

Comment (SteveSandersonMS) [reply src/Components/WebView/WebView/src/IpcSender.cs:33]: Looking at how `IpcCommon.Serialize` is implemented, we're missing **so many** opportunities to avoid allocations 😄 

 * It uses a `params` array which could be a specific overload for this perf-critical case (`OutgoingMessageType`, `long`, `string-or-whatever-we-choose`)
 * It calls `OutgoingMessageType.ToString()` without using any [cached lookup](https://www.meziantou.net/caching-enum-tostring-to-improve-performance.htm) (unless the compiler/runtime now does this automatically)
 * The `Serialize` method calls `array.Prepend`, allocating yet another array
 * The `Serialize` method calls `JsonSerializer.Serialize` to build a string, and then appends that to another string via interpolation, instead of doing the serialization directly into a single output buffer
 * We allocate a new `ArrayBuilder` *and* a new `ArrayBuilderMemoryStream` *and* a new `RenderBatchWriter` every single time, all of which could probably be reused or pooled. The large blocks of data representing the batch are passing through multiple representations through this process.

Basically, we didn't attempt any optimizations here at all, which is likely a good thing until we found a justification for it (premature optimizations and all that :)). But now we arguably do have some justification.

If we're really concerned about perf for how we build the IPC message, we should consider making an end-to-end optimized version of the call. For example, `IpcCommon` could have a new method `SerializeRenderBatchMessage` that takes the original `RenderBatch` and generates the final IPC message string as a single process into a single output buffer, possibly using `String.Create` but I'm not 100% certain on the detail without working through it.

**Review by javiercn — APPROVED** [review]

**Review by javiercn — APPROVED** [review]

File: src/Components/WebView/WebView/src/IpcSender.cs (hunk 2 above)
Comment (javiercn) [reply src/Components/WebView/WebView/src/IpcSender.cs:33]: @SteveSandersonMS  I have filed an issue to track the actual work we want to do [here](https://github.com/dotnet/maui/issues/12767)

**Comment by javiercn** [comment]
@softwaretirol on second look, I think we want to take the opportunity to step back and take a deeper look at the approach we use here rather than doing a one-off fix.

I've filed a separate issue [here](https://github.com/dotnet/maui/issues/12767) and we'll have a design discussion there. If you are interested in producing a proposal or want to contribute an implementation (after we've discussed a proposal) we will be happy to work with you.

For now, we are going to close this PR, although we might revisit in the future and merge it as a result of the discussion on the other issue.
